Rolling updater configurability enhancements

ironcladlou opened the pull request:
This PR adds some enhancements to the rolling updater:
1. Add an `UpdateAcceptor` interface to the rolling updater which supports injecting code to validate the first replica during scale-up. If the replica is not accepted, the deployment fails. This facilitates canary checking so that many broken replicas aren't rolled out during an update.
2. Make the rolling update scale amount configurable as a percent of the replica count.
3. Support scale directionality (down/up and up/down) to facilitate in-place deployments.

During the course of refactoring, I also changed the output of the updater as it progresses to something I consider more legible to an end-user. If the old output style is preferred, I can change it back.

ironcladlou committed:
Rolling updater enhancements

Add an UpdateAcceptor interface to the rolling updater which supports
injecting code to validate the first replica during scale-up. If the
replica is not accepted, the deployment fails. This facilitates canary
checking so that many broken replicas aren't rolled out during an update.

Make the rolling update scale amount configurable as a percent of the replica
count; a negative value changes the scale direction to down/up to support
in-place deployments.

ironcladlou commented:
cc @brendanburns @abhgupta @smarterclayton

ironcladlou commented:
For those interested in a use case, [here's an example in OpenShift](https://github.com/openshift/origin/blob/master/pkg/deploy/strategy/support/lifecycle.go#L242) of an `UpdateAcceptor` which waits for all containers in the replica/pod to be observed as ready before accepting the update.

ironcladlou commented:
I signed it!

smarterclayton commented:
This is post 1.0

smarterclayton commented:
ok to test

smarterclayton commented:
Or rather - this doesn't have to be in 1.0, it adds things that we use to implement what deployment will eventually need.  If it's not impactful I won't hate it being in 1.0

smarterclayton commented:
LGTM

ironcladlou commented:
This needs reviewed again since I've rolled in another enhancement which depends on the UpdateAcceptor.

ironcladlou commented:
>Ultimately it seems like we want to run the acceptor on all updates - why is the first one special? If acceptance depends on ready state, then every loop would want to wait on ready. It's more that we want to explicitly canary on the first loop (terminate early if we can't get any success).

Are you saying that the core code here should simply check for acceptance on every iteration, and a specific acceptance implementation could fail early if the first check fails if it wants?

smarterclayton commented:
Probably - if you don't want to compromise the deployment invariant (10% is
available) we have to wait for ready somehow

On Jul 6, 2015, at 11:34 AM, Dan Mace <notifications@github.com> wrote:

Ultimately it seems like we want to run the acceptor on all updates - why
is the first one special? If acceptance depends on ready state, then every
loop would want to wait on ready. It's more that we want to explicitly
canary on the first loop (terminate early if we can't get any success).

Are you saying that the core code here should simply check for acceptance
on every iteration, and a specific acceptance implementation could fail
early if the first check fails if it wants?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10062#issuecomment-118894870>
.

ironcladlou commented:
All feedback addressed.

ironcladlou commented:
@smarterclayton have time to review this or want to delegate? I'd like to start integrating in origin.

smarterclayton commented:
Maybe @derekwaynecarr or @kargakis can look while I'm on PTO this week

On Jul 7, 2015, at 1:59 PM, Dan Mace <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> have time to review
this or want to delegate? I'd like to start integrating in origin.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10062#issuecomment-119284222>
.

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -175,6 +207,21 @@ func UpdateExistingReplicationController(c client.Interface, oldRc *api.Replicat
+		if err != nil {
+			return nil, fmt.Errorf("Couldn't make scaler: %s", err)
+		}
+		if err := scaler.Scale(rc.Namespace, rc.Name, uint(rc.Spec.Replicas), &ScalePrecondition{-1, ""}, retry, wait); err != nil {
```
You dont need to specify any precondition here, the scaler will initialize it for you. I was wondering though if `wait` should be specified somewhere before you return the closure (I get it that we should wait for the scaling to finish).

kargakis replied:
Ok, got how this is working.

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,50 +400,83 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+		newSpec = &newRc.Spec
+
+		// Perform acceptance checking, if necessary.
+		if config.UpdateAcceptor != nil {
```
Shouldn't acceptance checking happen before scaling the controller?

kargakis replied:
I see it's on purpose after a scale-up.

kargakis commented on pkg/kubectl/rolling_updater_test.go:
```diff
@@ -132,165 +132,292 @@ func newRc(replicas int, desired int) *api.ReplicationController {
 		var buffer bytes.Buffer
+		var acceptor UpdateAcceptor
+		if test.accepted != nil {
+			acceptor = &testAcceptor{
```
Can you add a more realistic acceptance test than this? Make the acceptor part of the tests struct so it'll be more configurable and eg. try to accept a replication controlller with at most 2 replicas.

ironcladlou replied:
Not sure I follow what that adds since this code only cares about pass or fail from the acceptor, which I already exercise (via Accepted/Rejected booleans). Specific acceptor implementations elsewhere would have their own tests. Maybe I just misunderstood the scenario you're describing?

kargakis replied:
My understanding is that the acceptor will work based on a specific value in the status/spec of a replication controller. Or is it a specific type of acceptor I am describing?

ironcladlou replied:
>My understanding is that the acceptor will work based on a specific value in the status/spec of a replication controller. Or is it a specific type of acceptor I am describing?

You're describing a specific acceptor implementation. The rolling updater code only cares about passing the controller to the acceptor at the right time and either proceeding or aborting based on whether the black-box acceptor returns true or false.

kargakis replied:
:+1:

kargakis commented:
Looks good

General observation: Table tests without struct fields are so hard to follow.

ironcladlou commented:
@kargakis 
>Looks good
>General observation: Table tests without struct fields are so hard to follow.

I can add those.

ironcladlou commented:
@jlowdermilk @davidopp any thoughts here?

davidopp commented:
Since this is in the post-1.0 bucket, we probably won't be able to review it until next week. Sorry about the delay, and thanks for your patience!

ironcladlou commented:
@smarterclayton, bringing a side discussion over here. For supporting a configurable scaling direction to achieve scale-in-place, what do you think about another option like `ScaleOrder` which defaults to somthing like `UpFirst` and which can be set to `DownFirst`?

ironcladlou commented:
I also just realized that the current percentage implementation results in down/up asymmetry, which we explicitly didn't want. Shouldn't the scaling count be computed based on a percent of the desired replica count of the new controller, and that same value used for scaling up the new *and* scaling down the old? Right now I compute independent values for old/new,

smarterclayton commented:
I don't care either way - if negative and positive percentages aren't
obvious being explicit about direction as a secondary flag is fine.  I
don't know that there are more than two directions, and it's not really
direction so much as defining the behavior the user wants (use extra
capacity or use less capacity).  It's possible that if we can't scale up
due to capacity limitations we should scale down (otherwise out deployment
is wedged), in which case this is just a hint.

On Jul 15, 2015, at 12:42 PM, Dan Mace <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton>, bringing a side
discussion over here. For supporting a configurable scaling direction to
achieve scale-in-place, what do you think about another option like
ScaleOrder which defaults to somthing like UpFirst and which can be set to
DownFirst?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10062#issuecomment-121673748>
.

smarterclayton commented:
The total should correspond to the deployment invariant (desired total
pods).  If you add 10 pods because of 5%, you should remove 10 pods, not 5%
of the current size.

On Jul 15, 2015, at 12:45 PM, Dan Mace <notifications@github.com> wrote:

I also just realized that the current percentage implementation results in
down/up asymmetry, which we explicitly didn't want. Shouldn't the scaling
count be computed based on a percent of the desired replica count of the
new controller, and that same value used for scaling up the new *and*
scaling down the old? Right now I compute independent values for old/new,

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10062#issuecomment-121675196>
.

ironcladlou commented:
@smarterclayton I just completed a major refactor to the main rolling update loop. It now supports a change of scale direction based on whether the percentage is negative. I also fixed the scale amount symmetry for up/down. This refactor also handles all fencepost conditions uniformly without needing extra duplicated logic outside the loop as before. As a bonus to the fencepost fixes, update acceptance is applied uniformly on all scale up events (previously acceptance wasn't occuring for fencepost conditions). Output is also cleaned up to be more readable (IMO).

Fixed all the tests and added new ones to test direcitonality.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	// oldRc has 0 replicas.
+	for newSpec.Replicas != desired || oldSpec.Replicas != 0 {
+		// Set up a function for scaling up.
+		scaleUp := func() error {
```
This is an awkward construct - using a closure inside a loop isn't great. Some changes:

* The first block (setting newSpec.Replicas) can be abstracted into its own method that returns a bool.

ironcladlou replied:
Extracted this outside the loop and made it more stateless.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+			if err != nil {
+				return err
+			}
+			newSpec = &newRc.Spec
```
This is a bad idea to do in a closure.

ironcladlou replied:
Moved this outside the loop and made it more stateless.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+			}
+			newSpec = &newRc.Spec
+			// Perform acceptance checking, if necessary.
+			if config.UpdateAcceptor != nil {
```
You should have a no-op acceptor, and require the acceptor to print its own message.  Also, I'm not sure you really need to print out "Performing acceptance check".  That would reduce this to `if err := config.UpdateAcceptor.Accept(newRc); err != nil { return ... }`

ironcladlou replied:
Made these changes.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+				fmt.Fprintf(out, "Performing acceptance check of %s\n", newName)
+				err := config.UpdateAcceptor.Accept(newRc)
+				if err != nil {
+					return fmt.Errorf("Update rejected for %s: %v", newName, err)
```
Shouldn't be capitalized.

ironcladlou replied:
Fixed.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+
+		// Set up a function for scaling down.
+		scaleDown := func() error {
+			if oldSpec.Replicas == 0 {
```
This block can be extracted.

ironcladlou replied:
As above; extracted/made more stateless.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,49 +404,101 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
-			return err
+
+		// Perform the scaling interval in the specified direction.
+		if scaleUpFirst {
```
The difference between these two directions is `up, down, up, down` or `down, up, down, up`.  If ups and downs are no-ops when at the limits (which they should be), then you can do this loop like:

```
if !skipUpOnFirstLoop {
  ... scale up
  time.Sleep()
  skipUpOnFirstLoop = false
}
... scale down
```

smarterclayton replied:
There are better names for that variable - `skipFirstUp`

ironcladlou replied:
I implemented this. I'm not 100% convinced it's any clearer and it's only slightly smaller code, but I don't feel strongly, so if you like the way the new style looks, I have no objections.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,51 +411,94 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	// Set up a function for scaling up.
+	scaleUp := func(newRc, oldRc *api.ReplicationController) (*api.ReplicationController, error) {
+		// Nothing to do.
+		if newRc.Spec.Replicas == desired {
```
I think these could be split into their own functions - then you inline scaleUp in the block below.  I don't think you should use closures like this - it's not structurally their goal.

ironcladlou replied:
I thought it looked a little cleaner than making increasingly stateless functions which took a ton of params. Nevertheless...

Refactored to extract functions.

smarterclayton commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,51 +411,94 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		// Choose up/down vs. down/up scaling direction.
+		if !skipFirstUp {
+			if scaledRc, err := scaleUp(newRc, oldRc); err != nil {
```
```
rc, err := ....
if err != nil {
  return err
}
newRc = rc
```

ironcladlou replied:
Done

smarterclayton commented:
Looks good to me - this is fine for OpenShift and we can iterate further here post 1.0

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -353,51 +411,50 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
 		if err != nil {
 			return err
 		}
+		newRc = scaledRc
```
Shouldn't we `time.Sleep(updatePeriod)` after this scale-up?

ironcladlou replied:
I preserved the existing sleep behavior:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/rolling_updater.go#L357

mikedanese commented:
@ironcladlou can you rebase? I think you caught master in a bad state.

ironcladlou commented:
@mikedanese 
>@ironcladlou can you rebase? I think you caught master in a bad state.

Rebased.

mikedanese commented:
All green. @jlowdermilk @davidopp fyi
